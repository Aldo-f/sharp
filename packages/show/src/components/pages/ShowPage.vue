<template>
    <div class="ShowPage">
        <div class="container">
            <template v-if="ready">
                <ActionBarShow
                    :commands="authorizedCommands"
                    :state="instanceState"
                    :state-values="stateValues"
                    :form-url="formUrl"
                    :back-url="backUrl"
                    :can-edit="canEdit"
                    :can-change-state="canChangeState"
                    :show-back-button="showBackButton"
                    :breadcrumb="breadcrumbItems"
                    :show-breadcrumb="breadcrumb.visible"
                    @command="handleCommandRequested"
                    @state-change="handleStateChanged"
                />

                <div class="mt-4 pt-2">
                    <template v-for="section in layout.sections">
                        <Section
                            class="ShowPage__section"
                            v-show="isSectionVisible(section)"
                            :section="section"
                            :layout="sectionLayout(section)"
                            :fields-row-class="fieldsRowClass"
                            :collapsable="isSectionCollapsable(section)"
                            v-slot="{ fieldLayout }"
                        >
                            <template v-if="fieldOptions(fieldLayout)">
                                <ShowField
                                    :options="fieldOptions(fieldLayout)"
                                    :value="fieldValue(fieldLayout)"
                                    :config-identifier="fieldLayout.key"
                                    :layout="fieldLayout"
                                    :collapsable="section.collapsable"
                                    @visible-change="handleFieldVisibilityChanged(fieldLayout.key, $event)"
                                    :key="refreshKey"
                                />
                            </template>
                            <template v-else>
                                <UnknownField :name="fieldLayout.key" />
                            </template>
                        </Section>
                    </template>
                </div>
            </template>
            <template v-else>
                <ActionBarShow />
            </template>
        </div>

        <CommandFormModal :command="currentCommand" ref="commandForm" />
        <CommandViewPanel :content="commandViewContent" @close="handleCommandViewPanelClosed" />
    </div>
</template>

<script>
    import { mapGetters } from 'vuex';
    import { formUrl, getBackUrl, lang, showAlert, handleNotifications, withLoadingOverlay } from 'sharp';
    import { CommandFormModal, CommandViewPanel } from 'sharp-commands';
    import { Grid } from 'sharp-ui';
    import { UnknownField } from 'sharp/components';
    import { withCommands } from 'sharp/mixins';
    import ActionBarShow from "../ActionBar";

    import ShowField from '../Field';
    import Section from "../Section";

    export default {
        mixins: [withCommands],

        components: {
            Section,
            ActionBarShow,
            Grid,
            ShowField,
            UnknownField,
            CommandFormModal,
            CommandViewPanel,
        },

        data() {
            return {
                ready: false,
                fieldsVisible: null,
                refreshKey: 0,
            }
        },

        computed: {
            ...mapGetters('show', [
                'entityKey',
                'instanceId',
                'fields',
                'layout',
                'data',
                'config',
                'breadcrumb',
                'instanceState',
                'canEdit',
                'authorizedCommands',
                'stateValues',
                'canChangeState',
            ]),

            formUrl() {
                const formKey = this.config.multiformAttribute
                    ? this.data[this.config.multiformAttribute]
                    : null;
                const entityKey = formKey
                    ? `${this.entityKey}:${formKey}`
                    : this.entityKey

                return formUrl({
                    entityKey,
                    instanceId: this.instanceId,
                }, { append: true });
            },
            backUrl() {
                return getBackUrl(this.breadcrumb.items);
            },
            showBackButton() {
                return !!this.backUrl;
            },
            breadcrumbItems() {
                return this.breadcrumb.items;
            },
        },

        methods: {
            fieldOptions(layout) {
                const options = this.fields
                    ? this.fields[layout.key]
                    : null;
                if(!options) {
                    console.error(`Show page: unknown field "${layout.key}"`);
                }
                return options;
            },
            fieldValue(layout) {
                return this.data
                    ? this.data[layout.key]
                    : null;
            },
            isFieldVisible(layout) {
                return !this.fieldsVisible || this.fieldsVisible[layout.key] !== false;
            },
            isSectionCollapsable(section) {
                return section.collapsable && !this.sectionHasField(section, 'entityList');
            },
            sectionLayout(section) {
                if(this.sectionHasField(section, 'entityList')) {
                    return 'contents';
                }
                return 'card';
            },
            fieldsRowClass(row) {
                const fieldsTypeClasses = row.map(fieldLayout => {
                    const field = this.fieldOptions(fieldLayout);
                    return `ShowPage__fields-row--${field.type}`;
                });
                return [
                    'ShowPage__fields-row',
                    ...fieldsTypeClasses,
                ]
            },
            sectionFields(section) {
                return section.columns.reduce((res, column) => [...res, ...column.fields.flat()], []);
            },
            isSectionVisible(section) {
                const sectionFields = this.sectionFields(section);
                return sectionFields.some(fieldLayout => this.isFieldVisible(fieldLayout));
            },
            sectionHasField(section, type) {
                const sectionFields = this.sectionFields(section);
                return sectionFields.some(fieldLayout => {
                    const options = this.fieldOptions(fieldLayout);
                    return options && options.type === type;
                });
            },
            handleFieldVisibilityChanged(key, visible) {
                this.fieldsVisible = {
                    ...this.fieldsVisible,
                    [key]: visible,
                }
            },
            handleCommandRequested(command) {
                this.sendCommand(command, {
                    postCommand: () => this.$store.dispatch('show/postCommand', { command }),
                    postForm: data => this.$store.dispatch('show/postCommand', { command, data }),
                    getFormData: () => this.$store.dispatch('show/getCommandFormData', { command }),
                });
            },
            handleStateChanged(state) {
                return this.$store.dispatch('show/postState', state)
                    .then(data => {
                        this.handleCommandActionRequested(data.action, data);
                    })
                    .catch(error => {
                        const data = error.response.data;
                        if(error.response.status === 422) {
                            showAlert(data.message, {
                                title: lang('modals.state.422.title'),
                                isError: true,
                            });
                        }
                    });
            },
            handleRefreshCommand() {
                this.init();
            },
            initCommands() {
                this.addCommandActionHandlers({
                    'refresh': this.handleRefreshCommand,
                });
            },
            async init() {
                await this.$store.dispatch('show/setEntityKey', this.$route.params.entityKey);
                await this.$store.dispatch('show/setInstanceId', this.$route.params.instanceId);

                const show = await withLoadingOverlay(
                    this.$store.dispatch('show/get')
                        .catch(error => {
                            this.$emit('error', error);
                            return Promise.reject(error);
                        })
                );

                handleNotifications(show.notifications);

                this.ready = true;
                this.refreshKey++;
            }
        },

        beforeMount() {
            this.init();
            this.initCommands();
        },
    }
</script>
