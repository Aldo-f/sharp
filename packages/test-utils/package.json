{
    "name": "sharp-test-utils",
    "private": true
}