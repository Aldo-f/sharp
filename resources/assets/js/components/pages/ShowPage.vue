<template>
    <div class="SharpShowPage">
        <template v-if="ready">
            <SharpActionBarShow />
        </template>
    </div>
</template>

<script>
    import { mapState, mapGetters } from 'vuex';
    import SharpActionBarShow from "../action-bar/ActionBarShow";

    export default {
        components: {
            SharpActionBarShow,
        },

        data() {
            return {
                ready: false,
            }
        },

        computed: {
            ...mapState('show', {
                entityKey: state => state.entityKey,
            }),
            ...mapGetters('show', [
                'canEdit',
            ]),
        },

        methods: {
            async init() {
                await this.$store.dispatch('show/setEntityKey', this.$route.params.entityKey);
                await this.$store.dispatch('show/setInstanceId', this.$route.params.instanceId);
                await this.$store.dispatch('show/get');

                this.ready = true;
            }
        },

        created() {
            this.init();
        },
    }
</script>