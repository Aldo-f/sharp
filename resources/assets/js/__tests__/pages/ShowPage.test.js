import merge from 'lodash/merge';
import Vuex from 'vuex';
import ShowPage from "../../components/pages/ShowPage";
import { shallowMount, createLocalVue } from '@vue/test-utils';
import showModule from "../../store/modules/show";

jest.mock('../../store/modules/show');
jest.mock('../../util/url');


describe('show page', () => {
    function createWrapper({ storeModule, ...options } = {}) {
        const localVue = createLocalVue();
        localVue.use(Vuex);
        const wrapper = shallowMount(ShowPage, {
            extends: {
                computed: {
                    breadcrumb: () => [],
                    config: () => withDefaultConfig(),
                },
            },
            localVue,
            created() {
                jest.spyOn(this, 'init').mockImplementation(() => { });
            },
            store: new Vuex.Store({
                modules: {
                    'show': merge(showModule, storeModule),
                }
            }),
            stubs: {
                'SharpGrid': {
                    template: '<div class="MOCKED_SharpGrid" v-bind="$attrs"><slot v-bind="{}" /></div>',
                },
            },
            ...options,
        });
        return wrapper;
    }

    function withDefaultConfig(config) {
        return merge({
            state: {
                values: [],
            }
        }, config);
    }

    test('can mount', () => {
        expect(createWrapper().html()).toMatchSnapshot();
    });

    test('can mount with fields', () => {
        const wrapper = createWrapper({
            computed: {
                layout: () => ({
                    sections: [
                        {
                            title: 'Section title',
                            columns: []
                        }
                    ]
                }),
                fields: () => ({}),
                data: () => ({}),
            }
        });
        wrapper.setData({ ready: true });
        expect(wrapper.html()).toMatchSnapshot();
    });
});