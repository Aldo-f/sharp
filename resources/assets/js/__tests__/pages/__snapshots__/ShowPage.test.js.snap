// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show page can mount 1`] = `
<div class="ShowPage">
	<div class="container">
		<!---->
	</div>
	<sharpcommandformmodal-stub></sharpcommandformmodal-stub>
	<sharpcommandviewpanel-stub></sharpcommandviewpanel-stub>
</div>
`;

exports[`show page can mount with fields 1`] = `
<div class="ShowPage">
	<div class="container">
		<sharpactionbarshow-stub statevalues=""></sharpactionbarshow-stub>
		<h2 class="ShowPage__section-title mb-2">Section title</h2>
		<div class="ShowPage__section mb-4">
			<div rows="" class="MOCKED_SharpGrid">
				<div class="MOCKED_SharpGrid ShowPage__fields-grid">
					<sharpshowfield-stub></sharpshowfield-stub>
				</div>
			</div>
		</div>
	</div>
	<sharpcommandformmodal-stub></sharpcommandformmodal-stub>
	<sharpcommandviewpanel-stub></sharpcommandviewpanel-stub>
</div>
`;
